// @tarojs/taro-rn/dist/lib/connectSocket/socketTask.js
class SocketTask {
    constructor(url, protocols) {
        if (protocols) {
            // eslint-disable-next-line no-undef
            this.ws = new WebSocket(url, protocols);
        }
        else {
            // eslint-disable-next-line no-undef
            this.ws = new WebSocket(url);
        }
    }
    get CONNECTING() {
        return 0;
    }
    get OPEN() {
        return 1;
    }
    get CLOSING() {
        return 2;
    }
    get CLOSED() {
        return 3;
    }
    get readyState() {
        return this.ws.readyState;
    }
    send(opts) {
        const { data, success, fail, complete } = opts;
        const res = { errMsg: 'sendSocketMessage:ok' };
        try {
            this.ws.send(data);
            success === null || success === void 0 ? void 0 : success(res);
            complete === null || complete === void 0 ? void 0 : complete(res);
        }
        catch (err) {
            const res = { errMsg: err.message };
            fail === null || fail === void 0 ? void 0 : fail(res);
            complete === null || complete === void 0 ? void 0 : complete(res);
        }
    }
    close(opts) {
        const { code = 1000, reason = 'server complete, close', success, fail, complete } = opts || {};
        const res = { errMsg: 'closeSocket:ok' };
        this.closeDetail = { code, reason };
        try {
            this.ws.close(code, reason);
            // 把自己从链接数组中清除
            this._destroyWhenClose && this._destroyWhenClose();
            success === null || success === void 0 ? void 0 : success(res);
            complete === null || complete === void 0 ? void 0 : complete(res);
        }
        catch (err) {
            const res = { errMsg: err.message };
            fail === null || fail === void 0 ? void 0 : fail(res);
            complete === null || complete === void 0 ? void 0 : complete(res);
        }
    }
    onOpen(func) {
        this.ws.onopen = () => {
            func && func({
                header: {}
            });
        };
    }
    onMessage(func) {
        this.ws.onmessage = (res) => {
            func && func({
                data: res.data
            });
        };
    }
    onClose(func) {
        this.ws.onclose = () => {
            this._destroyWhenClose && this._destroyWhenClose();
            func && func(this.closeDetail || { code: 1006, reason: 'abnormal closure' });
        };
    }
    onError(func) {
        this.ws.onerror = () => {
            func && func({
                errMsg: "There was an error with your websocket.",
            });
        };
    }
}
export default SocketTask;
