// @tarojs/runtime-rn/dist/page.js
import { EventChannel } from '@tarojs/shared';
import { Component, createContext, createElement, createRef, forwardRef } from 'react';
import { AppState, Dimensions, RefreshControl, ScrollView } from 'react-native';
import { isClassComponent } from './app';
import { Current } from './current';
import { eventCenter } from './emmiter';
import { getCurrentRoute, isTabPage, PageProvider } from './router';
import { EMPTY_OBJ, errorHandler, getPageStr, incrementId, isArray, isFunction, successHandler } from './utils';
const compId = incrementId();
// 页面实例
const instances = new Map();
// 对标小程序的实例
const pagesObj = new Map();
function setPageObject(inst, id) {
    pagesObj.set(id, inst);
}
function getPageObject(id) {
    return pagesObj.get(id);
}
export function injectPageInstance(inst, id) {
    instances.set(id, inst);
}
export function getPageInstance(id) {
    return instances.get(id);
}
function getLifecyle(instance, lifecyle) {
    return instance[lifecyle];
}
function safeExecute(path, lifecycle, ...args) {
    const instance = instances.get(path);
    if (instance == null) {
        return;
    }
    const func = getLifecyle(instance, lifecycle);
    if (isArray(func)) {
        const res = func.map((fn) => fn.apply(instance, args));
        return res[0];
    }
    if (!isFunction(func)) {
        return;
    }
    return func.apply(instance, args);
}
const globalAny = global;
// eslint-disable-next-line import/no-mutable-exports
export let PageContext = EMPTY_OBJ;
// APP 前后台状态发生变化时调用对应的生命周期函数
let appState = AppState.currentState;
AppState.addEventListener('change', (nextAppState) => {
    const { page, app, router } = Current;
    if (!page)
        return;
    if (appState.match(/inactive|background/) && nextAppState === 'active') {
        (app === null || app === void 0 ? void 0 : app.onShow) &&
            app.onShow({
                path: router === null || router === void 0 ? void 0 : router.path,
                query: router === null || router === void 0 ? void 0 : router.params
            });
        if (!page.__isReactComponent && page.__safeExecute) {
            page.__safeExecute('componentDidShow');
        }
        else if (page.onShow) {
            page.onShow();
        }
    }
    if (appState === 'active' && nextAppState.match(/inactive|background/)) {
        if (!page.__isReactComponent && page.__safeExecute) {
            page.__safeExecute('componentDidHide');
        }
        else if (page.onHide) {
            page.onHide();
        }
        (app === null || app === void 0 ? void 0 : app.onHide) && app.onHide();
    }
    appState = nextAppState;
});
// 屏幕宽高发生变化
Dimensions.addEventListener('change', ({ window }) => {
    const { page } = Current;
    if (!page)
        return;
    if (!page.__isReactComponent && page.__safeExecute) {
        page.__safeExecute('onResize', { size: window });
    }
    else {
        if (page.onResize) {
            page.onResize({ size: window });
        }
    }
});
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function createPageConfig(Page, pageConfig) {
    const h = createElement;
    const pagePath = pageConfig.pagePath.replace(/^\//, '') || '';
    const isReactComponent = isClassComponent(Page);
    if (PageContext === EMPTY_OBJ) {
        PageContext = createContext('');
    }
    let ScreenPage = Page;
    if (!isReactComponent) {
        // eslint-disable-next-line react/display-name
        ScreenPage = forwardRef((props, ref) => {
            return h(Page, Object.assign({ forwardRef: ref }, props), null);
        });
    }
    const WrapScreen = (Screen) => {
        return class PageScreen extends Component {
            constructor(props) {
                var _a, _b, _c, _d;
                super(props);
                this.pullDownRefresh = (path, refresh) => {
                    if (getPageStr(path) === getPageStr(pagePath)) {
                        this.setState({ refreshing: refresh });
                    }
                };
                this.setRefreshStyle = () => {
                    var _a;
                    const refreshStyle = (_a = globalAny === null || globalAny === void 0 ? void 0 : globalAny.__taroRefreshStyle) !== null && _a !== void 0 ? _a : {};
                    this.setState({
                        textColor: refreshStyle.textColor || '#ffffff',
                        backgroundColor: refreshStyle.backgroundColor || '#ffffff'
                    });
                };
                this.pageToScroll = ({ path = '', scrollTop = 0 }) => {
                    var _a, _b;
                    if (getPageStr(path) === getPageStr(pagePath)) {
                        (_b = (_a = this.pageScrollView) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.scrollTo({ x: 0, y: scrollTop, animated: true });
                    }
                };
                const refreshStyle = (_a = globalAny === null || globalAny === void 0 ? void 0 : globalAny.__taroRefreshStyle) !== null && _a !== void 0 ? _a : {};
                const backgroundTextStyle = pageConfig.backgroundTextStyle || ((_d = (_c = (_b = globalAny.__taroAppConfig) === null || _b === void 0 ? void 0 : _b.appConfig) === null || _c === void 0 ? void 0 : _c.window) === null || _d === void 0 ? void 0 : _d.backgroundTextStyle) || 'dark';
                this.state = {
                    refreshing: false,
                    textColor: refreshStyle.textColor || (backgroundTextStyle === 'dark' ? '#000000' : '#ffffff'),
                    backgroundColor: refreshStyle.backgroundColor || '#ffffff'
                };
                appState = AppState.currentState;
                this.screenRef = createRef();
                this.pageScrollView = createRef();
                this.setPageInstance();
                this.pageId = `taro_page_${compId()}`;
            }
            componentDidMount() {
                var _a, _b, _c, _d, _e, _f;
                const { navigation, route } = this.props;
                // 实现 useLoad hook
                // handleHooksEvent 在组件构造函数中调用不生效，只能在挂载之后进行调用
                this.handleHooksEvent('onLoad', (_a = route === null || route === void 0 ? void 0 : route.params) !== null && _a !== void 0 ? _a : {});
                if (navigation) {
                    this.unSubscribleTabPress = navigation.addListener('tabPress', () => this.onTabItemTap());
                    this.unSubscribleFocus = navigation.addListener('focus', () => this.onFocusChange());
                    this.unSubscribleBlur = navigation.addListener('blur', () => this.onBlurChange());
                    // 如果是tabbar页面，因为tabbar是懒加载的，第一次点击事件还未监听，不会触发，初始化触发一下
                    const lazy = (_f = (_e = (_d = (_c = (_b = globalAny.__taroAppConfig) === null || _b === void 0 ? void 0 : _b.appConfig) === null || _c === void 0 ? void 0 : _c.rn) === null || _d === void 0 ? void 0 : _d.tabOptions) === null || _e === void 0 ? void 0 : _e.lazy) !== null && _f !== void 0 ? _f : true;
                    if (isTabPage() && lazy) {
                        this.onTabItemTap();
                    }
                }
                eventCenter.on('__taroPullDownRefresh', this.pullDownRefresh, this);
                eventCenter.on('__taroPageScrollTo', this.pageToScroll, this);
                eventCenter.on('__taroSetRefreshStyle', this.setRefreshStyle, this);
            }
            componentWillUnmount() {
                const { navigation, route } = this.props;
                eventCenter.off('__taroPullDownRefresh', this.pullDownRefresh, this);
                eventCenter.off('__taroPageScrollTo', this.pageToScroll, this);
                eventCenter.off('__taroSetRefreshStyle', this.setRefreshStyle, this);
                if (navigation) {
                    this.unSubscribleTabPress();
                    this.unSubscribleBlur();
                    this.unSubscribleFocus();
                }
                if (route && route.key) {
                    pagesObj.delete(route.key);
                }
                // 实现 useUnload hook
                this.handleHooksEvent('onUnload');
            }
            setPageInstance() {
                var _a;
                const pageRef = this.screenRef;
                const pageId = this.pageId;
                const { params = {}, key = '' } = (_a = this.props.route) !== null && _a !== void 0 ? _a : {};
                // 和小程序的page实例保持一致
                const inst = {
                    config: pageConfig,
                    route: pagePath,
                    options: params,
                    __isReactComponent: isReactComponent,
                    __safeExecute(lifecycle, ...rest) {
                        safeExecute(pageId, lifecycle, ...rest);
                    },
                    onReady() {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.componentDidMount)) {
                            page.componentDidMount && page.componentDidMount();
                        }
                        else {
                            safeExecute(pageId, 'onReady');
                        }
                    },
                    onShow() {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.componentDidShow)) {
                            page.componentDidShow && page.componentDidShow();
                        }
                        else {
                            safeExecute(pageId, 'componentDidShow');
                        }
                    },
                    onHide() {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.componentDidHide)) {
                            page.componentDidHide && page.componentDidHide();
                        }
                        else {
                            safeExecute(pageId, 'componentDidHide');
                        }
                    },
                    onPullDownRefresh() {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.onPullDownRefresh)) {
                            page.onPullDownRefresh && page.onPullDownRefresh();
                        }
                        else {
                            safeExecute(pageId, 'onPullDownRefresh');
                        }
                    },
                    onReachBottom() {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.onReachBottom)) {
                            page.onReachBottom && page.onReachBottom();
                        }
                        else {
                            safeExecute(pageId, 'onReachBottom');
                        }
                    },
                    onPageScroll(options) {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.onPageScroll)) {
                            page.onPageScroll && page.onPageScroll(options);
                        }
                        else {
                            safeExecute(pageId, 'onPageScroll', options);
                        }
                    },
                    onResize(options) {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.onResize)) {
                            page.onResize && page.onResize(options);
                        }
                        else {
                            safeExecute(pageId, 'onResize', options);
                        }
                    },
                    onTabItemTap(options) {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.onTabItemTap)) {
                            page.onTabItemTap && page.onTabItemTap(options);
                        }
                        else {
                            safeExecute(pageId, 'onTabItemTap', options);
                        }
                    },
                    onUnload() {
                        const page = pageRef.current;
                        if (page != null && isFunction(page.componentWillUnmount)) {
                            page.componentWillUnmount && page.componentWillUnmount();
                        }
                        else {
                            safeExecute(pageId, 'onUnload');
                        }
                    },
                    getOpenerEventChannel() {
                        return EventChannel.pageChannel;
                    }
                };
                // 存储对应小程序的实例
                setPageObject(inst, key);
                Current.router = {
                    params: params,
                    path: pagePath.startsWith('/') ? pagePath : `/${pagePath}`
                };
                Current.page = inst;
            }
            onFocusChange() {
                var _a, _b, _c;
                // 页面切换，当前instance重新赋值
                this.setPageInstance();
                try {
                    this.handleHooksEvent('componentDidShow');
                    // 实现 useReady hook，遵循小程序事件机制，在useDidShow之后触发
                    if (!this.isPageReady) {
                        this.handleHooksEvent('onReady');
                        this.isPageReady = true;
                    }
                    (_c = (_b = (_a = this.screenRef) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.componentDidShow) === null || _c === void 0 ? void 0 : _c.call(_b);
                }
                catch (err) {
                    throw new Error(err);
                }
            }
            onBlurChange() {
                var _a, _b, _c, _d;
                try {
                    this.handleHooksEvent('componentDidHide');
                    if ((_b = (_a = this.screenRef) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.componentDidHide) {
                        (_d = (_c = this.screenRef) === null || _c === void 0 ? void 0 : _c.current) === null || _d === void 0 ? void 0 : _d.componentDidHide();
                    }
                }
                catch (err) {
                    throw new Error(err);
                }
            }
            onPageScroll(e) {
                var _a, _b, _c, _d;
                if (!(e === null || e === void 0 ? void 0 : e.nativeEvent))
                    return;
                const { contentOffset } = e.nativeEvent;
                const scrollTop = contentOffset.y;
                if (scrollTop < 0)
                    return;
                try {
                    this.handleHooksEvent('onPageScroll', { scrollTop });
                    if ((_b = (_a = this.screenRef) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.onPageScroll) {
                        (_d = (_c = this.screenRef) === null || _c === void 0 ? void 0 : _c.current) === null || _d === void 0 ? void 0 : _d.onPageScroll({ scrollTop });
                    }
                }
                catch (err) {
                    throw new Error(err);
                }
            }
            // 监听的onMomentumScrollEnd
            onReachBottom(e) {
                var _a, _b, _c, _d;
                if (!(e === null || e === void 0 ? void 0 : e.nativeEvent))
                    return;
                const { onReachBottomDistance = 50 } = pageConfig;
                const { layoutMeasurement, contentSize, contentOffset } = e.nativeEvent;
                if ((contentOffset === null || contentOffset === void 0 ? void 0 : contentOffset.y) + (layoutMeasurement === null || layoutMeasurement === void 0 ? void 0 : layoutMeasurement.height) + onReachBottomDistance >= contentSize.height) {
                    try {
                        this.handleHooksEvent('onReachBottom');
                        if ((_b = (_a = this.screenRef) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.onReachBottom) {
                            (_d = (_c = this.screenRef) === null || _c === void 0 ? void 0 : _c.current) === null || _d === void 0 ? void 0 : _d.onReachBottom();
                        }
                    }
                    catch (err) {
                        throw new Error(err);
                    }
                }
            }
            // 下拉刷新
            onPullDownRefresh() {
                var _a, _b, _c, _d;
                this.setState({ refreshing: true });
                try {
                    this.handleHooksEvent('onPullDownRefresh');
                    if ((_b = (_a = this.screenRef) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.onPullDownRefresh) {
                        (_d = (_c = this.screenRef) === null || _c === void 0 ? void 0 : _c.current) === null || _d === void 0 ? void 0 : _d.onPullDownRefresh();
                    }
                }
                catch (e) {
                    throw new Error(e);
                }
                finally {
                    this.setState({ refreshing: false });
                }
            }
            onTabItemTap() {
                var _a, _b, _c, _d;
                try {
                    const item = this.getTabItem(pagePath);
                    this.handleHooksEvent('onTabItemTap', Object.assign({}, item));
                    if ((_b = (_a = this.screenRef) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.onTabItemTap) {
                        (_d = (_c = this.screenRef) === null || _c === void 0 ? void 0 : _c.current) === null || _d === void 0 ? void 0 : _d.onTabItemTap(item);
                    }
                }
                catch (error) {
                    throw new Error(error);
                }
            }
            handleHooksEvent(method, options = {}) {
                return safeExecute(this.pageId, method, options);
            }
            getTabItem(itemPath) {
                var _a, _b;
                const tabBar = ((_b = (_a = globalAny.__taroAppConfig) === null || _a === void 0 ? void 0 : _a.appConfig) === null || _b === void 0 ? void 0 : _b.tabBar) || {};
                if (!tabBar)
                    return '';
                let result = {};
                for (let i = 0; i < tabBar.list.length; i++) {
                    const item = tabBar.list[i];
                    const path = item.pagePath.replace(/^\//, '') || '';
                    if (getPageStr(path) === getPageStr(itemPath)) {
                        result = {
                            index: i,
                            pagePath: path,
                            text: item.text
                        };
                    }
                }
                return result;
            }
            isEnablePullDown() {
                var _a, _b;
                const windowOptions = ((_b = (_a = globalAny.__taroAppConfig) === null || _a === void 0 ? void 0 : _a.appConfig) === null || _b === void 0 ? void 0 : _b.window) || {};
                const { enablePullDownRefresh = false } = windowOptions;
                return (pageConfig === null || pageConfig === void 0 ? void 0 : pageConfig.enablePullDownRefresh) || enablePullDownRefresh;
            }
            refreshPullDown() {
                const { refreshing, textColor, backgroundColor } = this.state;
                return createElement(RefreshControl, {
                    refreshing: refreshing,
                    enabled: true,
                    titleColor: textColor,
                    tintColor: textColor,
                    colors: [backgroundColor],
                    onRefresh: () => this.onPullDownRefresh()
                }, null);
            }
            createPage() {
                if (PageProvider) {
                    return h(PageProvider, Object.assign({ currentPath: pagePath, pageConfig }, this.props), h(PageContext.Provider, { value: this.pageId }, h(Screen, Object.assign(Object.assign({}, this.props), { ref: this.screenRef }))));
                }
                return h(PageContext.Provider, { value: this.pageId }, h(Screen, Object.assign(Object.assign({}, this.props), { ref: this.screenRef })));
            }
            createScrollPage() {
                var _a, _b, _c, _d, _e;
                let bgColor = pageConfig.backgroundColor ? pageConfig.backgroundColor : '';
                const windowOptions = ((_b = (_a = globalAny.__taroAppConfig) === null || _a === void 0 ? void 0 : _a.appConfig) === null || _b === void 0 ? void 0 : _b.window) || {};
                const useNativeStack = (_e = (_d = (_c = globalAny.__taroAppConfig) === null || _c === void 0 ? void 0 : _c.appConfig) === null || _d === void 0 ? void 0 : _d.rn) === null || _e === void 0 ? void 0 : _e.useNativeStack;
                if (!bgColor && (windowOptions === null || windowOptions === void 0 ? void 0 : windowOptions.backgroundColor)) {
                    bgColor = windowOptions === null || windowOptions === void 0 ? void 0 : windowOptions.backgroundColor;
                }
                const refresh = this.isEnablePullDown() ? { refreshControl: this.refreshPullDown() } : {};
                return h(ScrollView, Object.assign(Object.assign({ style: [{ flex: 1 }, bgColor ? { backgroundColor: bgColor } : {}], contentContainerStyle: useNativeStack ? {} : { minHeight: '100%' }, ref: this.pageScrollView, scrollEventThrottle: 8 }, refresh), { onScroll: (e) => this.onPageScroll(e), onMomentumScrollEnd: (e) => this.onReachBottom(e), nestedScrollEnabled: true }), this.createPage());
            }
            render() {
                const { disableScroll = false } = pageConfig;
                return !disableScroll ? this.createScrollPage() : this.createPage();
            }
        };
    };
    const pageComponet = WrapScreen(ScreenPage);
    return pageComponet;
}
export function startPullDownRefresh(options = {}) {
    const currentPage = Current.page;
    const path = currentPage === null || currentPage === void 0 ? void 0 : currentPage.route;
    const { success, fail, complete } = options;
    let errMsg = 'startPullDownRefresh:ok';
    try {
        eventCenter.trigger('__taroPullDownRefresh', { path, refresh: true });
        success === null || success === void 0 ? void 0 : success({ errMsg });
        complete === null || complete === void 0 ? void 0 : complete({ errMsg });
        return Promise.resolve({ errMsg });
    }
    catch (error) {
        errMsg = 'startPullDownRefresh:fail';
        fail === null || fail === void 0 ? void 0 : fail({ errMsg });
        complete === null || complete === void 0 ? void 0 : complete({ errMsg });
        return Promise.reject(error);
    }
}
export function stopPullDownRefresh(options = {}) {
    const currentPage = Current.page;
    const path = currentPage === null || currentPage === void 0 ? void 0 : currentPage.route;
    const { success, fail, complete } = options;
    let errMsg = 'stopPullDownRefresh:ok';
    try {
        eventCenter.trigger('__taroPullDownRefresh', { path, refresh: false });
        success === null || success === void 0 ? void 0 : success({ errMsg });
        complete === null || complete === void 0 ? void 0 : complete({ errMsg });
        return Promise.resolve({ errMsg });
    }
    catch (error) {
        errMsg = 'stopPullDownRefresh:fail';
        fail === null || fail === void 0 ? void 0 : fail({ errMsg });
        complete === null || complete === void 0 ? void 0 : complete({ errMsg });
        return Promise.reject(error);
    }
}
export function pageScrollTo(options = {}) {
    const currentPage = Current.page;
    const path = currentPage === null || currentPage === void 0 ? void 0 : currentPage.route;
    const { success, fail, complete, scrollTop = 0 } = options;
    let errMsg = 'pageScrollTo:ok';
    try {
        eventCenter.trigger('__taroPageScrollTo', { path, scrollTop });
        success === null || success === void 0 ? void 0 : success({ errMsg });
        complete === null || complete === void 0 ? void 0 : complete({ errMsg });
        return Promise.resolve({ errMsg });
    }
    catch (error) {
        errMsg = 'pageScrollTo:fail';
        fail === null || fail === void 0 ? void 0 : fail({ errMsg });
        complete === null || complete === void 0 ? void 0 : complete({ errMsg });
        return Promise.reject(error);
    }
}
// 仅支持android
export function setBackgroundColor(options) {
    const { backgroundColor, success, fail, complete } = options;
    const errMsg = ' setBackgroundColor: ok';
    const refreshStyle = globalAny.__taroRefreshStyle || {};
    try {
        refreshStyle.backgroundColor = backgroundColor;
        globalAny.__taroRefreshStyle = refreshStyle;
        eventCenter.trigger('__taroSetRefreshStyle');
        return successHandler(success, complete)({ errMsg });
    }
    catch (error) {
        const errMsg = ' setBackgroundColor: error';
        return errorHandler(fail, complete)({ errMsg });
    }
}
// 仅支持ios
export function setBackgroundTextStyle(options) {
    const { textStyle, success, fail, complete } = options;
    const textColor = textStyle === 'dark' ? '#000000' : '#ffffff';
    const errMsg = ' setBackgroundTextStyle: ok';
    const refreshStyle = globalAny.__taroRefreshStyle || {};
    try {
        refreshStyle.textColor = textColor;
        globalAny.__taroRefreshStyle = refreshStyle;
        eventCenter.trigger('__taroSetRefreshStyle');
        return successHandler(success, complete)({ errMsg });
    }
    catch (error) {
        const errMsg = ' setBackgroundTextStyle: error';
        return errorHandler(fail, complete)({ errMsg });
    }
}
export function getCurrentPages() {
    const pages = [];
    const routes = getCurrentRoute();
    if (routes && routes.length > 0) {
        routes.forEach((item) => {
            const inst = getPageObject(item);
            inst && pages.push(inst);
        });
    }
    else {
        // 第一次初始化时，getCurrentRoute会为空
        const inst = Current.page;
        inst && pages.push(inst);
    }
    return pages;
}
