// @tarojs/components-rn/dist/components/Slider/index.js
/**
 * 半受控组件
 *
 * ✔ min
 * ✔ max
 * ✔ step
 * ✔ disabled
 * ✔ value
 * - color
 * - selected-color
 * ✔ activeColor
 * ✔ backgroundColor
 * ✘ blockSize(block-size)
 * ✔ blockColor(block-color)
 * ✔ showValue(show-value)
 * ✔ onChange(bindchange)
 * ✔ onChanging(bindchanging)
 */
import * as React from 'react';
import { View, Text } from 'react-native';
import Slider from '@react-native-community/slider';
import { noop } from '../../utils';
import styles from './styles';
class _Slider extends React.Component {
    constructor() {
        var _a;
        super(...arguments);
        this._timer = null;
        // eslint-disable-next-line eqeqeq
        this._isControlled = this.props.value != undefined;
        this.state = {
            currentValue: (_a = this.props.defaultValue) !== null && _a !== void 0 ? _a : (this.props.value || 0)
        };
        this.ensureControlledValue = () => {
            if (!this._isControlled)
                return;
            this._timer && clearTimeout(this._timer);
            this._timer = setTimeout(() => {
                const controlledValue = this.props.value;
                if (controlledValue !== this.state.currentValue) {
                    this.setState({
                        currentValue: controlledValue || 0
                    });
                }
            }, 50);
        };
        this.onSlidingComplete = (value) => {
            const { onChange = noop } = this.props;
            onChange({ detail: { value } });
        };
        this.onValueChange = (value) => {
            const { onChanging = noop } = this.props;
            onChanging({ detail: { value } });
            this.setState({ currentValue: value });
        };
    }
    componentDidUpdate(prevProps) {
        const controlledValue = this.props.value;
        if (prevProps.value !== controlledValue) {
            this.setState({
                currentValue: controlledValue || 0
            });
        }
        this.ensureControlledValue();
    }
    render() {
        const { style, min, max, step, disabled, activeColor, backgroundColor, blockColor, showValue } = this.props;
        // @todo dismember style
        return (React.createElement(View, { style: styles.wrapper },
            React.createElement(Slider, { minimumValue: min, maximumValue: max, step: step, disabled: !!disabled, value: this.state.currentValue, minimumTrackTintColor: activeColor, maximumTrackTintColor: backgroundColor, thumbTintColor: blockColor, onSlidingComplete: this.onSlidingComplete, onValueChange: this.onValueChange, style: [styles.bar, style] }),
            showValue && React.createElement(Text, { style: styles.info }, this.state.currentValue)));
    }
}
_Slider.displayName = '_Slider';
_Slider.defaultProps = {
    min: 0,
    max: 100,
    step: 1,
    activeColor: '#1aad19',
    backgroundColor: '#e9e9e9',
    blockColor: '#fff'
};
export default _Slider;
