// @tarojs/components-rn/dist/components/Radio/index.js
/**
 * 半受控组件
 *
 * ✔ value
 * ✔ disabled
 * ✔ checked
 * ✔ color
 * ✔ onChange
 */
import * as React from 'react';
import { TouchableWithoutFeedback } from 'react-native';
import Icon from '../Icon';
import View from '../View';
import styles from './styles';
import { noop } from '../../utils';
class _Radio extends React.Component {
    constructor() {
        super(...arguments);
        this.$touchable = React.createRef();
        this.state = {
            checked: false
        };
        this._simulateNativePress = (evt) => {
            const node = this.$touchable.current;
            node && node.props.onPress && node.props.onPress(evt);
        };
        this.onPress = () => {
            const { disabled, onChange = noop, value } = this.props;
            if (disabled)
                return;
            if (this.state.checked)
                return;
            onChange({
                value,
                checked: !this.state.checked
            });
            this.setState({ checked: !this.state.checked });
        };
    }
    static getDerivedStateFromProps(props, state) {
        return props.checked !== state.checked
            ? {
                checked: !!props.checked
            }
            : null;
    }
    render() {
        const { style, color } = this.props;
        const isChecked = this.state.checked;
        return (React.createElement(TouchableWithoutFeedback, { onPress: this.onPress, ref: this.$touchable },
            React.createElement(View, { style: styles.container },
                React.createElement(View, { style: [styles.wrapper, isChecked && styles.wrapperChecked, isChecked && { borderColor: color }, style] },
                    React.createElement(Icon, { type: "success", size: 24, color: color, style: [styles.wrapperIcon, isChecked && styles.wrapperCheckedIcon] })),
                React.createElement(View, { style: { flexGrow: 0 } }, this.props.children))));
    }
}
_Radio.displayName = '_Radio';
_Radio.defaultProps = {
    value: '',
    color: '#09BB07'
};
export default _Radio;
