// @tarojs/components-rn/dist/components/CheckboxGroup/index.js
/**
 * ✔ onChange(bindchange)
 *
 * @warn No support of props FOR, you must put checkbox below label straightly.
 */
import * as React from 'react';
import { View, } from 'react-native';
import { noop } from '../../utils';
class _CheckboxGroup extends React.Component {
    constructor() {
        super(...arguments);
        this.values = [];
        this.tmpIndex = 0;
        this.getDataFromValues = () => {
            return this.values
                .filter((item) => item.checked)
                .map((item) => item.value);
        };
        this.toggleChange = (e, index) => {
            const { onChange = noop } = this.props;
            this.values[index] = {
                value: e.value,
                checked: e.checked
            };
            onChange({
                detail: {
                    value: this.getDataFromValues()
                }
            });
        };
        this.findAndAttachCb = (children) => {
            return React.Children.toArray(children).map((child) => {
                if (!child.type)
                    return child;
                const childTypeName = child.type.displayName;
                if (childTypeName === '_Checkbox') {
                    const { value, disabled, checked, color } = child.props;
                    const index = this.tmpIndex++;
                    this.values[index] = { value, checked };
                    return React.cloneElement(child, {
                        onChange: (e) => this.toggleChange(e, index),
                        value,
                        disabled,
                        checked,
                        color
                    });
                }
                else {
                    return React.cloneElement(child, Object.assign({}, child.props), this.findAndAttachCb(child.props.children));
                }
            });
        };
    }
    render() {
        const { children, style, _onGroupDataInitial = noop } = this.props;
        this.tmpIndex = 0;
        const mapChildren = this.findAndAttachCb(children);
        _onGroupDataInitial(this.getDataFromValues());
        return (React.createElement(View, { style: style }, mapChildren));
    }
}
_CheckboxGroup.displayName = '_CheckboxGroup';
export default _CheckboxGroup;
