// @tarojs/taro-h5/dist/api/open-api/index.js
export { getAccountInfoSync } from './account.js';
export { chooseAddress } from './address.js';
export { authorize, authorizeForMiniProgram } from './authorize.js';
export { addCard, openCard } from './card.js';
export { getChannelsLiveInfo, getChannelsLiveNoticeInfo, getChannelsShareKey, openChannelsActivity, openChannelsEvent, openChannelsLive, openChannelsUserProfile, reserveChannelsLive } from './channels-live.js';
export { openCustomerServiceChat } from './customer-service.js';
export { getDeviceVoIPList, requestDeviceVoIP } from './device-voip.js';
export { checkIsSupportFacialRecognition, faceVerifyForPay, startFacialRecognitionVerify, startFacialRecognitionVerifyAndUploadVideo } from './facial.js';
export { addFileToFavorites, addVideoToFavorites } from './favorites.js';
export { getGroupEnterInfo } from './group.js';
export { chooseInvoice, chooseInvoiceTitle } from './invoice.js';
export { chooseLicensePlate } from './license-plate.js';
export { checkSession, login, pluginLogin } from './login.js';
export { checkIsAddedToMyMiniProgram } from './my-miniprogram.js';
export { getPrivacySetting, onNeedPrivacyAuthorization, openPrivacyContract, requirePrivacyAuthorize } from './privacy.js';
export { showRedPackage } from './red-package.js';
export { getSetting, openSetting } from './settings.js';
export { checkIsSoterEnrolledInDevice, checkIsSupportSoterAuthentication, startSoterAuthentication } from './soter.js';
export { requestSubscribeDeviceMessage, requestSubscribeMessage } from './subscribe-message.js';
export { getUserInfo, getUserProfile } from './user-info.js';
export { getWeRunData, shareToWeRun } from './werun.js';
