// @tarojs/components-rn/dist/components/Checkbox/index.js
/**
 * 半受控组件
 *
 * ✔ value
 * ✔ disabled
 * ✔ checked
 * ✔ color
 * ✔ onChange
 *
 * @see https://wechat.design/brand/color
 */
import * as React from 'react';
import { TouchableWithoutFeedback } from 'react-native';
import Icon from '../Icon';
import View from '../View';
import styles from './styles';
import { noop } from '../../utils';
class _Checkbox extends React.Component {
    constructor() {
        super(...arguments);
        this.$touchable = React.createRef();
        this.state = {
            checked: false,
        };
        this._simulateNativePress = (evt) => {
            const node = this.$touchable.current;
            node && node.props.onPress && node.props.onPress(evt);
        };
        this.onPress = () => {
            const { disabled, onChange = noop, value } = this.props;
            if (disabled)
                return;
            const checked = !this.state.checked;
            onChange({
                value,
                checked
            });
            this.setState({ checked });
        };
    }
    static getDerivedStateFromProps(props, state) {
        return props.checked !== state.checked
            ? {
                checked: !!props.checked
            }
            : null;
    }
    render() {
        const { style, color } = this.props;
        const { checked } = this.state;
        return (React.createElement(TouchableWithoutFeedback, { onPress: this.onPress, ref: this.$touchable },
            React.createElement(View, { style: styles.container },
                React.createElement(View, { style: [styles.wrapper, style, checked && styles.wrapperChecked] },
                    React.createElement(Icon, { type: "success_no_circle", size: 18, color: color, style: [styles.wrapperIcon, checked && styles.wrapperCheckedIcon] })),
                React.createElement(View, { style: { flexGrow: 0 } }, this.props.children))));
    }
}
_Checkbox.displayName = '_Checkbox';
_Checkbox.defaultProps = {
    value: '',
    color: '#09BB07'
};
export default _Checkbox;
