// @tarojs/taro-rn/dist/lib/createVideoContext/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const globalAny = global;
globalAny._taroVideoMap = globalAny._taroVideoMap || {};
class VideoContext {
    constructor(videoRef) {
        this.videoRef = videoRef;
    }
    /**
     * 退出全屏
     */
    exitFullScreen() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield ((_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.dismissFullscreenPlayer());
            }
            catch (e) {
                console.log(e);
            }
        });
    }
    /**
     * 隐藏状态栏，仅在iOS全屏下有效
     * @todo
     */
    hideStatusBar() {
        console.log('not support');
    }
    /**
     * 暂停视频
     */
    pause() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield ((_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.pauseAsync());
            }
            catch (e) {
                console.log(e);
            }
        });
    }
    /**
     * 播放视频
     */
    play() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield ((_b = (_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.playAsync) === null || _b === void 0 ? void 0 : _b.call(_a));
            }
            catch (e) {
                console.log(e);
            }
        });
    }
    /**
     * 设置倍速播放
     * {number} @param rate - 倍率，支持 0.5/0.8/1.0/1.25/1.5，2.6.3 起支持 2.0 倍速
     */
    playbackRate(rate) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield ((_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.setRateAsync(rate));
            }
            catch (e) {
                console.log(e);
            }
        });
    }
    /**
     * 进入全屏
     * @package {object} [object]
     * @package {number} [object.direction] - 设置全屏时视频的方向，不指定则根据宽高比自动判断。
     */
    requestFullScreen() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield ((_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.presentFullscreenPlayer());
            }
            catch (e) {
                console.log(e);
            }
        });
    }
    /**
     * 跳转到指定位置
     * @param {number} position - 跳转到的位置，单位 s
     */
    seek(position) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const millis = position * 1000;
            try {
                yield ((_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.setPositionAsync(millis));
            }
            catch (e) {
                console.log(e);
            }
        });
    }
    /**
     * 发送弹幕 ❌
     * @ todo
     * @deprecated 暂未实现
     * @param {DanmuData} data 弹幕内容
     * @param {string} data.text 弹幕文字
     * @param {string} data.color 弹幕颜色
     */
    sendDanmu() {
        console.log('not support');
    }
    /**
     * 显示状态栏，仅在iOS全屏下有效
     * @todo
     */
    showStatusBar() {
        console.log('not support');
    }
    exitBackgroundPlayback() {
        console.log('not support');
    }
    exitPictureInPicture() {
        console.log('not support');
    }
    requestBackgroundPlayback() {
        console.log('not support');
    }
    /**
     * 停止视频
     */
    stop() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield ((_a = this.videoRef) === null || _a === void 0 ? void 0 : _a.stopAsync());
            }
            catch (e) {
                console.log(e);
            }
        });
    }
}
/**
 * 创建 video 上下文 VideoContext 对象。
 * {string} @param - id video 组件的 id
 * {object} @param t - 在自定义组件下，当前组件实例的this，以操作组件内 video 组件
 */
export function createVideoContext(id, t) {
    let ref = globalAny._taroVideoMap[id];
    if (t)
        ref = t;
    if (ref) {
        return new VideoContext(ref);
    }
    else {
        return undefined;
    }
}
