// @tarojs/router-rn/dist/utils/index.js
import { camelCase } from 'lodash';
import { parseUrl } from 'query-string';
const globalAny = global;
export function isUrl(str) {
    const protocolAndDomainRE = /^(?:\w+:)?\/\/(\S+)$/;
    const localhostDomainRE = /^localhost[:?\d]*(?:[^:?\d]\S*)?$/;
    const nonLocalhostDomainRE = /^[^\s.]+\.\S{2,}$/;
    if (typeof str !== 'string') {
        return false;
    }
    const match = str.match(protocolAndDomainRE);
    if (!match) {
        return false;
    }
    const everythingAfterProtocol = match[1];
    if (!everythingAfterProtocol) {
        return false;
    }
    if (localhostDomainRE.test(everythingAfterProtocol) ||
        nonLocalhostDomainRE.test(everythingAfterProtocol)) {
        return true;
    }
    return false;
}
export function isFunction(o) {
    return typeof o === 'function';
}
export function isEmptyObject(obj) {
    if (obj == null) {
        return true;
    }
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            return false;
        }
    }
    return true;
}
export function successHandler(success, complete) {
    return function (res) {
        success && isFunction(success) && success(res);
        complete && isFunction(complete) && complete(res);
        return Promise.resolve(res);
    };
}
export function errorHandler(fail, complete) {
    return function (res) {
        fail && isFunction(fail) && fail(res);
        complete && isFunction(complete) && complete(res);
        return Promise.reject(res);
    };
}
export function getTabItemConfig(index, key) {
    const _taroTabBarIconConfig = globalAny.__taroTabBarIconConfig;
    const _taroTabItems = _taroTabBarIconConfig.tabItems;
    return _taroTabItems[index] && _taroTabItems[index][key];
}
export function getTabConfig(key) {
    const _taroTabBarIconConfig = globalAny.__taroTabBarIconConfig;
    return _taroTabBarIconConfig[key];
}
export function setTabConfig(key, value) {
    const tabBarConfig = globalAny.__taroTabBarIconConfig;
    tabBarConfig[key] = value;
    globalAny.__taroTabBarIconConfig = tabBarConfig;
}
export function setTabInitRoute(routeName) {
    globalAny.__taroTabInitRoute = routeName;
}
export function getTabInitRoute() {
    return globalAny.__taroTabInitRoute || '';
}
export function getTabVisible() {
    return getTabConfig('tabBarVisible');
}
export function getDefaultTabItem(index) {
    var _a;
    const _taroAppConfig = globalAny.__taroAppConfig || {};
    const tabBar = ((_a = _taroAppConfig === null || _taroAppConfig === void 0 ? void 0 : _taroAppConfig.appConfig) === null || _a === void 0 ? void 0 : _a.tabBar) || [];
    return (tabBar === null || tabBar === void 0 ? void 0 : tabBar.list[index]) || {};
}
// camelCase之后的页面
export function getTabBarPages() {
    var _a, _b;
    const tabBar = ((_b = (_a = globalAny.__taroAppConfig) === null || _a === void 0 ? void 0 : _a.appConfig) === null || _b === void 0 ? void 0 : _b.tabBar) || {};
    if (isEmptyObject(tabBar))
        return [];
    const pages = [];
    tabBar === null || tabBar === void 0 ? void 0 : tabBar.list.forEach((item) => {
        const path = item.pagePath.startsWith('/') ? item.pagePath : `/${item.pagePath}`;
        pages.push(camelCase(path));
    });
    return pages;
}
// 处理url转换成pageName与params
export function handleUrl(url) {
    const path = url.split('?')[0];
    const pageName = camelCase(path.startsWith('/') ? path : `/${path}`);
    const params = parseUrl(url.startsWith('/') ? url.substr(1) : url).query || {};
    return {
        pageName,
        params
    };
}
export function hasJumpAnimate() {
    if (globalAny.__taroJumpAnimate === false) {
        return false;
    }
    return true;
}
export function updateJumpAnimate(needAnimate) {
    globalAny.__taroJumpAnimate = needAnimate;
}
export function updateCurrentJumpUrl(path) {
    globalAny.__taroJumpUrl = path;
}
export function getCurrentJumpUrl() {
    var _a;
    return (_a = globalAny === null || globalAny === void 0 ? void 0 : globalAny.__taroJumpUrl) !== null && _a !== void 0 ? _a : '';
}
