// @tarojs/plugin-platform-tt/dist/components-react.js
const View = 'view';
const Icon = 'icon';
const Progress = 'progress';
const RichText = 'rich-text';
const Text = 'text';
const Button = 'button';
const Checkbox = 'checkbox';
const CheckboxGroup = 'checkbox-group';
const Form = 'form';
const Input = 'input';
const Label = 'label';
const Picker = 'picker';
const PickerView = 'picker-view';
const PickerViewColumn = 'picker-view-column';
const Radio = 'radio';
const RadioGroup = 'radio-group';
const Slider = 'slider';
const Switch = 'switch';
const CoverImage = 'cover-image';
const Textarea = 'textarea';
const CoverView = 'cover-view';
const MovableArea = 'movable-area';
const MovableView = 'movable-view';
const ScrollView = 'scroll-view';
const Swiper = 'swiper';
const SwiperItem = 'swiper-item';
const Navigator = 'navigator';
const Audio = 'audio';
const Camera = 'camera';
const Image = 'image';
const LivePlayer = 'live-player';
const Video = 'video';
const Canvas = 'canvas';
const Ad = 'ad';
const WebView = 'web-view';
const Block = 'block';
const Map = 'map';
const Slot = 'slot';
const NativeSlot = 'native-slot';
const CustomWrapper = 'custom-wrapper';

const RtcRoom = 'rtc-room';
const PayButton = 'pay-button';
const ConsumeCard = 'consume-card';
const AwemeData = 'aweme-data';
const RateButton = 'rate-button';
const OpenData = 'open-data';
const AwemeUserCard = 'aweme-user-card';

export { Ad, Audio, AwemeData, AwemeUserCard, Block, Button, Camera, Canvas, Checkbox, CheckboxGroup, ConsumeCard, CoverImage, CoverView, CustomWrapper, Form, Icon, Image, Input, Label, LivePlayer, Map, MovableArea, MovableView, NativeSlot, Navigator, OpenData, PayButton, Picker, PickerView, PickerViewColumn, Progress, Radio, RadioGroup, RateButton, RichText, RtcRoom, ScrollView, Slider, Slot, Swiper, SwiperItem, Switch, Text, Textarea, Video, View, WebView };
