// @tarojs/runtime/dist/index.js
import './dom-external/index.js';
export { default as env } from './env.js';
export { Events, hooks } from '@tarojs/shared';
export { document } from './bom/document.js';
export { getComputedStyle } from './bom/getComputedStyle.js';
export { History } from './bom/history.js';
export { Location } from './bom/location.js';
export { nav as navigator } from './bom/navigator.js';
export { caf as cancelAnimationFrame, now, raf as requestAnimationFrame } from './bom/raf.js';
export { URL, parseUrl } from './bom/URL.js';
export { URLSearchParams } from './bom/URLSearchParams.js';
export { history, location, window } from './bom/window.js';
export { TaroElement } from './dom/element.js';
export { TaroEvent, createEvent, eventHandler } from './dom/event.js';
export { FormElement } from './dom/form.js';
export { TaroNode } from './dom/node.js';
export { TaroRootElement } from './dom/root.js';
export { Style } from './dom/style.js';
export { SVGElement } from './dom/svg.js';
export { TaroText } from './dom/text.js';
export { MutationObserver } from './dom-external/mutation-observer/index.js';
export { A, APP, BEHAVIORS, BODY, CATCHMOVE, CATCH_VIEW, CHANGE, CLASS, COMMENT, COMPILE_MODE, CONFIRM, CONTAINER, CONTEXT_ACTIONS, CURRENT_TARGET, CUSTOM_WRAPPER, DATASET, DATE, DOCUMENT_ELEMENT_NAME, DOCUMENT_FRAGMENT, EVENT_CALLBACK_RESULT, EXTERNAL_CLASSES, FOCUS, HEAD, HOOKS_APP_ID, HTML, ID, INPUT, KEY_CODE, OBJECT, ON_HIDE, ON_LOAD, ON_READY, ON_SHOW, OPTIONS, PAGE_INIT, PROPERTY_THRESHOLD, PROPS, PURE_VIEW, ROOT_STR, SET_DATA, SET_TIMEOUT, STATIC_VIEW, STYLE, TARGET, TARO_RUNTIME, TIME_STAMP, TOUCHMOVE, TYPE, UID, VALUE, VIEW } from './constants/index.js';
export { Current, getCurrentInstance } from './current.js';
export { eventSource } from './dom/event-source.js';
export { createComponentConfig, createPageConfig, createRecursiveComponentConfig, getOnHideEventKey, getOnReadyEventKey, getOnShowEventKey, getPageInstance, getPath, injectPageInstance, removePageInstance, safeExecute, stringify } from './dsl/common.js';
export { eventCenter } from './emitter/emitter.js';
export { hydrate } from './hydrate.js';
export { nextTick } from './next-tick.js';
export { options } from './options.js';
export { incrementId, isHasExtractProp } from './utils/index.js';
export { addLeadingSlash, getCurrentPage, getHomePage, hasBasename, stripBasename, stripSuffix, stripTrailing } from './utils/router.js';
export { handlePolyfill } from './polyfill/index.js';
