// @tarojs/components-rn/dist/components/Camera/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import React, { Component } from 'react';
import View from '../View';
import Text from '../Text';
import { Camera, PermissionStatus } from 'expo-camera';
import { BarCodeScanner } from 'expo-barcode-scanner';
import styles from './styles';
export class _Camera extends Component {
    constructor(props) {
        super(props);
        this.expoCameraRef = React.createRef();
        this.onError = (event) => {
            this.props.onError && this.props.onError(event);
        };
        this.onInitDone = () => {
            global._taroCamera = this.expoCameraRef && this.expoCameraRef.current;
            const event = {};
            this.props.onInitDone && this.props.onInitDone(event);
        };
        this.onScanCode = (event) => {
            const { data } = event;
            this.props.onScanCode && this.props.onScanCode(Object.assign({ detail: {
                    result: data
                } }, event));
        };
        this.state = {
            hasPermission: null,
        };
    }
    componentDidMount() {
        return __awaiter(this, void 0, void 0, function* () {
            const { status } = yield Camera.requestCameraPermissionsAsync();
            this.setState({
                hasPermission: status === PermissionStatus.GRANTED
            });
        });
    }
    render() {
        const { hasPermission } = this.state;
        const { devicePosition, style, mode, flash } = this.props;
        const type = !devicePosition ? null : Camera.Constants.Type[devicePosition];
        if (hasPermission === null) {
            return React.createElement(View, null);
        }
        if (hasPermission === false) {
            return React.createElement(Text, null, "No access to camera");
        }
        const barCodeScannerSettings = mode === 'scanCode'
            ? {
                barCodeScannerSettings: {
                    barCodeTypes: [BarCodeScanner.Constants.BarCodeType.qr]
                },
                onBarCodeScanned: this.onScanCode
            }
            : {};
        return (React.createElement(Camera, Object.assign({ ref: this.expoCameraRef, 
            // @ts-ignore
            type: type, flashMode: flash, onMountError: this.onError, onCameraReady: this.onInitDone, ratio: this.props.ratio }, barCodeScannerSettings, { style: [styles.camera, style] })));
    }
}
export default _Camera;
